### ISSUE TITLE
Shallow Train

### ISSUE THREAD

**Issue by connerohnesorge** [connerohnesorge/sddec25-01#120]
Performance Optimization Plan for train.py

 Summary

 Optimize the ShallowDenseNet2D training pipeline by addressing CPU-bound preprocessing, unnecessary device transfers, missing PyTorch optimizations, and inefficient I/O patterns.

 Identified Bottlenecks (Priority Order)

 1. CPU-Bound Preprocessing in Dataset (CRITICAL)

 Location: IrisDataset.__getitem__ (lines 577-630)

 Issues:
 - one_hot2dist() distance transform computed on every batch access - O(N²) scipy operation
 - Canny edge detection (cv2.Canny) computed per-sample
 - CLAHE applied per-sample instead of cached
 - Augmentations applied sequentially

 Fix: Precompute distance maps and spatial weights during dataset initialization, cache CLAHE results.

 2. Unnecessary CPU-GPU Transfers (HIGH)

 Location: Training loop (lines 922-956) and validation loop (lines 987-1011)

 Issues:
 - spatialWeights.to(device) called per-batch (line 934)
 - maxDist.to(device) called per-batch (line 938)
 - mIoU() calls .item() multiple times per batch (line 131) - forces GPU sync

 Fix:
 - DataLoader's collate_fn should handle device placement
 - Accumulate IoU tensors, compute mean at epoch end
 - Use non_blocking=True for all .to(device) calls

 3. Missing PyTorch 2.0 Optimizations (HIGH)

 Location: Model initialization and training loop

 Issues:
 - No torch.compile() - missing 20-40% speedup on modern GPUs
 - No Automatic Mixed Precision (AMP) - T4 has tensor cores
 - No gradient checkpointing consideration

 Fix: Add torch.compile(), enable torch.cuda.amp.autocast(), use GradScaler

 4. Redundant Computations in Loss Functions (MEDIUM)

 Location: Loss classes (lines 70-109)

 Issues:
 - SurfaceLoss.forward() computes torch.softmax() (line 76)
 - GeneralizedDiceLoss.forward() also normalizes with softmax (line 91)
 - Model output gets softmax applied twice when both losses are used

 Fix: Compute softmax once, pass normalized tensor to both losses

 5. Memory-Inefficient Model Forward Pass (MEDIUM)

 Location: ShallowDenseNet2D.forward() (lines 477-493)

 Issues:
 - Intermediate activations stored as instance attributes (self.x1, self.x2, etc.)
 - These persist beyond forward pass, consuming GPU memory
 - Prevents garbage collection during training

 Fix: Use local variables instead of instance attributes

 6. Excessive I/O During Training (LOW)

 Location: Checkpoint/logging logic (lines 1066-1093)

 Issues:
 - ONNX export every 2 epochs is slow (line 1089)
 - Matplotlib plots generated synchronously
 - Best model exported on every improvement

 Fix: Export checkpoints less frequently, use background threads for I/O

 ---
 Implementation Plan

 Phase 1: Quick Wins (No Architecture Changes)

 1. Add torch.compile(model) after model initialization
 2. Wrap training loop with torch.cuda.amp.autocast() and GradScaler
 3. Add non_blocking=True to all .to(device) calls
 4. Replace .item() calls with tensor accumulation in mIoU()

 Phase 2: Data Loading Optimization

 1. Precompute distance maps and spatial weights in IrisDataset.__init__
 2. Store precomputed tensors in memory-mapped files or cache
 3. Move augmentations to GPU using torchvision.transforms.v2 or Kornia
 4. Consider prefetch_factor in DataLoader

 Phase 3: Loss Function Optimization

 1. Create unified CombinedLoss class that computes softmax once
 2. Pass pre-normalized tensor to all loss components
 3. Fuse boundary-aware weighting into single operation

 Phase 4: Model Memory Optimization

 1. Replace self.x1, self.x2, etc. with local variables
 2. Consider torch.utils.checkpoint for memory-compute tradeoff

 Phase 5: I/O Optimization

 1. Reduce checkpoint frequency to every 5 epochs
 2. Move ONNX export to background thread
 3. Batch MLflow logging calls

 ---
 Files to Modify

 - /home/connerohnesorge/Documents/001Repos/KRIA-UNET-PIPELINE/scripts/py/train_shallow/train.py

 Expected Performance Gains

 - AMP + torch.compile: 30-50% training speedup
 - Precomputed distance maps: 2-5x data loading speedup
 - Reduced device transfers: 10-15% per-batch speedup
 - Memory optimization: ~20% GPU memory reduction

 Testing Strategy

 1. Run baseline training for 5 epochs, record time per epoch
 2. Apply optimizations incrementally, measure after each phase
 3. Verify model accuracy is unchanged (mIoU within 0.5%)
 4. Profile with torch.profiler to validate bottleneck removal

 ---
 Detailed Implementation

 Change 1: Add Mixed Precision Training (AMP)

 # After model initialization (line ~718)
 scaler = torch.amp.GradScaler('cuda')

 # Wrap forward pass in training loop
 with torch.amp.autocast('cuda'):
     output = model(data)
     # loss computations...

 # Replace loss.backward() + optimizer.step()
 scaler.scale(total_loss).backward()
 scaler.step(optimizer)
 scaler.update()
 optimizer.zero_grad(set_to_none=True)  # More efficient than zero_grad()

 Change 2: Add torch.compile()

 # After model initialization (line ~718)
 model = torch.compile(model, mode='reduce-overhead')

 Change 3: Precompute Distance Maps in Dataset

 class IrisDataset(Dataset):
     def __init__(self, filepath, split='train', transform=None):
         # ... existing init code ...

         # Precompute all distance maps and spatial weights
         self.precomputed = {}
         if split != 'test':
             print(f"Precomputing distance maps for {split} set...")
             for idx, filename in enumerate(tqdm(self.list_files)):
                 labelpath = os.path.join(self.filepath, 'labels', filename + '.npy')
                 label = np.load(labelpath)
                 label = np.resize(label, (W, H))

                 # Map to binary
                 label_binary = np.zeros_like(label)
                 label_binary[label == 3] = 1

                 # Precompute spatial weights
                 spatialWeights = cv2.Canny(label_binary.astype(np.uint8) * 255, 0, 3) / 255
                 spatialWeights = cv2.dilate(spatialWeights, (3, 3), iterations=1) * 20

                 # Precompute distance maps
                 distMap = np.stack([one_hot2dist(label_binary == i) for i in range(2)], 0)

                 self.precomputed[filename] = {
                     'spatial_weights': spatialWeights.astype(np.float32),
                     'dist_map': distMap.astype(np.float32),
                     'label': label_binary.astype(np.uint8)
                 }

 Change 4: Accumulate IoU Without CPU Sync

 def mIoU_accumulated(predictions, targets, num_classes=2):
     """Compute IoU without CPU transfer - accumulate intersection/union tensors."""
     intersection = torch.zeros(num_classes, device=predictions.device)
     union = torch.zeros(num_classes, device=predictions.device)

     for c in range(num_classes):
         pred_c = predictions == c
         target_c = targets == c
         intersection[c] = torch.logical_and(pred_c, target_c).sum()
         union[c] = torch.logical_or(pred_c, target_c).sum()

     return intersection, union

 # At epoch end:
 # iou = (total_intersection / total_union.clamp(min=1)).mean().item()

 Change 5: Fused CombinedLoss Class

 class CombinedLoss(nn.Module):
     """Fused loss that computes softmax once for all components."""
     def __init__(self, epsilon=1e-5):
         super().__init__()
         self.epsilon = epsilon
         self.nll = nn.NLLLoss(reduction='none')

     def forward(self, logits, target, spatial_weights, dist_map, alpha):
         # Single softmax computation
         probs = F.softmax(logits, dim=1)
         log_probs = F.log_softmax(logits, dim=1)

         # Cross-entropy with spatial weighting
         ce_loss = self.nll(log_probs, target)
         weighted_ce = (ce_loss * (1 + spatial_weights)).mean()

         # Generalized Dice Loss (using pre-computed softmax)
         target_onehot = F.one_hot(target, num_classes=2).permute(0, 3, 1, 2).float()
         probs_flat = probs.flatten(start_dim=2)
         target_flat = target_onehot.flatten(start_dim=2)

         intersection = (probs_flat * target_flat).sum(dim=2)
         cardinality = (probs_flat + target_flat).sum(dim=2)
         class_weights = 1.0 / (target_flat.sum(dim=2) ** 2).clamp(min=self.epsilon)

         dice = 2 * (class_weights * intersection).sum(dim=1) / (class_weights * cardinality).sum(dim=1)
         dice_loss = (1 - dice.clamp(min=self.epsilon)).mean()

         # Surface Loss (using pre-computed softmax)
         surface_loss = (probs.flatten(start_dim=2) * dist_map.flatten(start_dim=2)).mean()

         # Combined loss
         total = weighted_ce + alpha * dice_loss + (1 - alpha) * surface_loss
         return total, weighted_ce, dice_loss, surface_loss

 Change 6: Model Memory Optimization

 def forward(self, x):
     # Use local variables instead of instance attributes
     x1 = self.down_block1(x)
     x2 = self.down_block2(x1)
     x3 = self.down_block3(x2)
     x4 = self.down_block4(x3)

     x5 = self.up_block1(x3, x4)
     x6 = self.up_block2(x2, x5)
     x7 = self.up_block3(x1, x6)

     if self.dropout:
         out = self.out_conv1(self.dropout1(x7))
     else:
         out = self.out_conv1(x7)
     return out

 Change 7: Non-blocking Transfers and Optimized DataLoader

 # DataLoader with prefetching
 trainloader = DataLoader(
     train_dataset,
     batch_size=BATCH_SIZE,
     shuffle=True,
     num_workers=NUM_WORKERS,
     pin_memory=True,
     persistent_workers=True,
     prefetch_factor=2,
     drop_last=True  # Avoids variable batch size overhead
 )

 # In training loop - non-blocking transfers
 data = img.to(device, non_blocking=True)
 target = labels.to(device, non_blocking=True)
 spatial_weights = spatialWeights.to(device, non_blocking=True)
 dist_map = maxDist.to(device, non_blocking=True)

 Change 8: Reduce Checkpoint Frequency

 # Change from every 2 epochs to every 10 epochs
 if (epoch + 1) % 10 == 0 or epoch == EPOCHS - 1:
     checkpoint_path = f"shallow_model_epoch_{epoch+1}.onnx"
     # ...

 ---
 Implementation Order (Each as Separate Commit)

 1. Commit 1: Add AMP + torch.compile() + non-blocking transfers
 2. Commit 2: Precompute distance maps/spatial weights in Dataset.init
 3. Commit 3: Create CombinedLoss class with single softmax
 4. Commit 4: Fix model forward() to use local variables
 5. Commit 5: Add accumulated IoU metric
 6. Commit 6: Reduce checkpoint frequency, optimize DataLoader settings

 Estimated Impact

 | Optimization     | Speed Gain        | Memory Reduction |
 |------------------|-------------------|------------------|
 | AMP              | 25-35%            | 30-40%           |
 | torch.compile    | 15-25%            | 0%               |
 | Precomputed data | 3-5x data loading | N/A              |
 | Fused loss       | 10-15%            | 5%               |
 | Local variables  | 0%                | 15-20%           |
 | Non-blocking     | 5-10%             | 0%               |

 Total expected: 50-80% faster training, 40-50% memory reduction
